doc/design/upgrade.md

```diff
@@ -125,15 +135,114 @@ Here is the upgrade workflow
 - Cleanup env after upgrade (remove upgrade snapshot, remove uploaded iso files if upgrade failed)
 
 This ensures that a failure to upload ISOs for whatever reason is detected before we take down the existing version. It limits the failure modes after shutting down the appliance to:
+
 1. failed to update appliance configuration
 2. failed to power on the appliance (e.g. system resource constraints)
 3. failed to boot appliance
 4. failed to acquire network addresses (this has been seen in real world examples - we may wish to attempt to preserve/reuse the IPs the appliance had prior to update, which should still be present in the extraconfig)
 5. failed to rebuild system state from vsphere infrastructure
 
+## Design - Phase2
+
+Although we did lots of code refactor for VCH appliance configuration, it's still unavoidable to continue changing that structure. To make sure we don't break  upgrade after GA, we need a solution for the data migration anyway.
+
+- Appliance guestinfo, which is used to persist VCH appliance configuration
+- Container guestinfo, which is used to persist container configuration
+- KeyValue Store, which is used to persist image metadata and network portlayer data
+- Appliance log files
+- vSphere object management logic change
+
+### Guestinfo Migration
+
+- One VirtualContainerHost structure per one version, this will be used by extraconfig package to read back old vch's configuration
+- Write plugin to migrate data from from previous version, and register that plugin with correct release version
+- Each version should have a separate directory, which contains all configuration files and plugins for that version
+- vic-machine upgrade framework will calculate the plugin chain based on old VCH's version and vic-machine version, and then call those plugins sequentially. This will look like r1->r2->r3->f4->latest or r3->r4->r5
+- Run guestinfo validation to figure out if there is any field missing (one possible option here is to translate configuration back to vic-machine input data, and then rerun validator to make sure all vsphere related objects are still exist and correct)
+- Prompt missing parameters if there is any (for new parameters added and cannot be empty)
+- Write new configuration back to appliance guestinfo
+
+With this design, developer is able to add/delete/repurpose existing configuration items, except VCH version. And we'll need one guidline for what to do while one configuration item is changed.
+
+Besides of this solution, we also dicussed another option, that is to revert configuration back to vic-machine create option, and try to recreate a new VCH appliance, and then migrate containers from the old appliance to new appliance. This solution can reduce VCH downtime, and even migrate alive.
```

Comment:
For here, I have some different idea as I mentioned in the VCH reconfigure issue. There have pros and cons for different options. Let's talk about that in person first.